File: DailyNews/src/main/java/com/jw/dailyNews/adapter/JokeAdapter.java
package com.jw.dailyNews.adapter;

import android.content.Context;
import android.graphics.Color;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.jw.dailyNews.R;
import com.jw.dailyNews.bean.Joke;
import com.jw.dailyNews.utils.DateUtils;
import com.jw.dailyNews.utils.GlideUtils;
import com.jw.dailyNews.wiget.ImageDetailDialog;
import com.jw.dailyNews.wiget.ImageLoadingPage;

import java.util.ArrayList;
import java.util.List;

import Lib.NewsManager;
import butterknife.BindView;
import butterknife.ButterKnife;
import de.hdodenhof.circleimageview.CircleImageView;
import fm.jiecao.jcvideoplayer_lib.JCVideoPlayer;
import fm.jiecao.jcvideoplayer_lib.JCVideoPlayerStandard;

public class JokeAdapter extends DefaultAdapter<Joke> {

    public JokeAdapter(Context context, List<Joke> lists) {
        super(context, lists);
    }

    public enum ITEM_TYPE {
        ITEM_TYPE_TEXT,
        ITEM_TYPE_IMAGE,
        ITEM_TYPE_GIF,
        ITEM_TYPE_VIDEO
    }


    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == ITEM_TYPE.ITEM_TYPE_IMAGE.ordinal()) {
            return new ImageViewHolder(mInflater.inflate(R.layout.item_joke_image, parent, false));
        } else if (viewType == ITEM_TYPE.ITEM_TYPE_TEXT.ordinal()) {
            return new TextViewHolder(mInflater.inflate(R.layout.item_joke_text, parent, false));
        } else if (viewType == ITEM_TYPE.ITEM_TYPE_GIF.ordinal()) {
            return new GifViewHolder(mInflater.inflate(R.layout.item_joke_gif, parent, false));
        } else if (viewType == ITEM_TYPE.ITEM_TYPE_VIDEO.ordinal()) {
            return new VideoViewHolder(mInflater.inflate(R.layout.item_joke_vidio, parent, false));
        } else {
            return null;
        }
    }

    @Override
    public void onBindViewHolder(final RecyclerView.ViewHolder holder, int position) {
        final Joke joke = lists.get(position);
        if (holder instanceof TextViewHolder) {

        } else if (holder instanceof ImageViewHolder) {
            GlideUtils.load(mContext,joke.getImage().getBig().get(0),((ImageViewHolder) holder).ivImage);
            ((ImageViewHolder) holder).btnLarge.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    ImageDetailDialog dialog = new ImageDetailDialog(mContext, joke.getImage().getBig().get(0));
                    dialog.show();
                }
            });
        } else if (holder instanceof GifViewHolder) {
            String url=joke.getGif().getImages().get(0);
            Log.v("gifurl",url);
            ImageLoadingPage imageLoadingPage=new ImageLoadingPage(mContext);
            imageLoadingPage.setLayoutParams(new FrameLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));
            ((GifViewHolder) holder).lll.removeAllViews();
            ((GifViewHolder) holder).lll.addView(imageLoadingPage);
            ((GifViewHolder) holder).lll.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    ImageDetailDialog dialog = new ImageDetailDialog(mContext, joke.getGif().getImages().get(0));
                    dialog.show();
                }
            });
            imageLoadingPage.setImageUrl(url);

        } else if (holder instanceof VideoViewHolder) {
            ((VideoViewHolder) holder).jcPlayer.setUp(joke.getVideo().getVideo().get(0), JCVideoPlayer.SCREEN_LAYOUT_LIST, joke.getText());
            GlideUtils.load(mContext,joke.getVideo().getThumbnail().get(0),((VideoViewHolder) holder).jcPlayer.thumbImageView);
        }
        GlideUtils.load(mContext,joke.getU().getHeader().get(0),((BaseHolder) holder).ivIconPublisher);
        ((BaseHolder) holder).tvNamePublisher.setText(joke.getU().getName());
        ((BaseHolder) holder).tvTimePublish.setText(DateUtils.fromNow("yyyy-MM-dd HH:mm:ss", joke.getPasstime()));
        ((BaseHolder) holder).tvTitle.setText(joke.getText());
        ((BaseHolder) holder).tvUp.setText(joke.getUp());
        ((BaseHolder) holder).tvDown.setText(joke.getDown() + "");
        ((BaseHolder) holder).tvForward.setText(joke.getForwrard() + "");
        ((BaseHolder) holder).tvCommentCount.setText(joke.getComment() + "");
        ((BaseHolder) holder).llForward.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                NewsManager.getInstance().showShare(joke.getText(),"",
                joke.getShare_url());
            }
        });
        ArrayList<Joke.TopComments> top_comments = joke.getTop_comments();
        if (top_comments != null && top_comments.size() != 0) {
            ((BaseHolder) holder).topComments.removeAllViews();
            for (Joke.TopComments comment : top_comments) {
                //行
                LinearLayout linearLayout = new LinearLayout(mContext);
                LinearLayout.LayoutParams layoutParams1 = new LinearLayout.LayoutParams(
                        LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT);
                LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(
                        LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT);
                linearLayout.setLayoutParams(layoutParams1);
                linearLayout.setOrientation(LinearLayout.HORIZONTAL);
                linearLayout.setPadding(5, 2, 5, 2);

                //用户名
                TextView tv_name = new TextView(mContext);
                tv_name.setLayoutParams(layoutParams);
                tv_name.setText(comment.getU().getName() + ": ");
                tv_name.setTextColor(Color.parseColor("#ff33b5e5"));

                //评论内容
                TextView tv_content = new TextView(mContext);
                tv_name.setLayoutParams(layoutParams);
                tv_content.setText(comment.getContent());
                tv_content.setTextColor(Color.BLACK);

                linearLayout.addView(tv_name);
                linearLayout.addView(tv_content);
                linearLayout.setBackgroundColor(Color.parseColor("#ffd7d7d7"));
                ((BaseHolder) holder).topComments.addView(linearLayout);
            }
        }
    }

    @Override
    public int getItemViewType(int position) {
        int ordinal = 0;
        switch (lists.get(position).getType()) {
            case "text":
                ordinal = ITEM_TYPE.ITEM_TYPE_TEXT.ordinal();   //序号
                break;
            case "image":
                ordinal = ITEM_TYPE.ITEM_TYPE_IMAGE.ordinal();
                break;
            case "gif":
                ordinal = ITEM_TYPE.ITEM_TYPE_GIF.ordinal();
                break;
            case "video":
                ordinal = ITEM_TYPE.ITEM_TYPE_VIDEO.ordinal();
        }
        return ordinal;
    }

    @Override
    public int getItemCount() {
        return lists == null ? 0 : lists.size();
    }


    public static class TextViewHolder extends BaseHolder {

        public TextViewHolder(View itemView) {
            super(itemView);
        }
    }

    public static class ImageViewHolder extends BaseHolder {
        @BindView(R.id.iv_image)
        ImageView ivImage;
        @BindView(R.id.btn_large)
        TextView btnLarge;

        public ImageViewHolder(View itemView) {
            super(itemView);
        }
    }

    public static class GifViewHolder extends BaseHolder {
        @BindView(R.id.llll)
        LinearLayout lll;

        public GifViewHolder(View itemView) {
            super(itemView);
        }
    }

    public static class VideoViewHolder extends BaseHolder {
        @BindView(R.id.jc_player)
        public JCVideoPlayerStandard jcPlayer;

        public VideoViewHolder(View itemView) {
            super(itemView);
        }
    }

    public static class BaseHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.iv_icon_publisher)
        CircleImageView ivIconPublisher;
        @BindView(R.id.tv_name_publisher)
        TextView tvNamePublisher;
        @BindView(R.id.tv_time_publish)
        TextView tvTimePublish;
        @BindView(R.id.tv_title)
        TextView tvTitle;

        @BindView(R.id.ib_up)
        ImageButton ibUp;
        @BindView(R.id.tv_up)
        TextView tvUp;
        @BindView(R.id.ib_down)
        ImageButton ibDown;
        @BindView(R.id.tv_down)
        TextView tvDown;
        @BindView(R.id.ib_forward)
        ImageView ibForward;
        @BindView(R.id.tv_forward)
        TextView tvForward;
        @BindView(R.id.ib_comment)
        ImageButton ibComment;
        @BindView(R.id.tv_commentCount)
        TextView tvCommentCount;
        @BindView(R.id.ll_forward)
        LinearLayout llForward;

        @BindView(R.id.top_comments)
        LinearLayout topComments;

        public BaseHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

}

File: dailyNews/src/main/java/com/jw/dailyNews/adapter/JokeAdapter.kt
package com.jw.dailyNews.adapter

import Lib.NewsManager
import android.content.Context
import android.databinding.DataBindingUtil
import android.databinding.ViewDataBinding
import android.graphics.Color
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.widget.*
import com.android.databinding.library.baseAdapters.BR
import com.jw.dailyNews.R
import com.jw.dailyNews.bean.Joke
import com.jw.dailyNews.databinding.ItemJokeGifBinding
import com.jw.dailyNews.databinding.ItemJokeImageBinding
import com.jw.dailyNews.databinding.ItemJokeTextBinding
import com.jw.dailyNews.databinding.ItemJokeVidioBinding
import com.jw.dailyNews.utils.DateUtils
import com.jw.dailyNews.utils.GlideUtils
import com.jw.dailyNews.wiget.ImageDetailDialog
import com.jw.dailyNews.wiget.ImageLoadingPage
import de.hdodenhof.circleimageview.CircleImageView
import fm.jiecao.jcvideoplayer_lib.JCVideoPlayer
import fm.jiecao.jcvideoplayer_lib.JCVideoPlayerStandard

class JokeAdapter(context: Context, lists: List<Joke>) : DefaultAdapter<Joke>(context, lists) {

    enum class ITEM_TYPE {
        ITEM_TYPE_TEXT,
        ITEM_TYPE_IMAGE,
        ITEM_TYPE_GIF,
        ITEM_TYPE_VIDEO
    }


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder? {
        return if (viewType == ITEM_TYPE.ITEM_TYPE_IMAGE.ordinal) {
            ImageViewHolder(mInflater.inflate(R.layout.item_joke_image, parent, false))
        } else if (viewType == ITEM_TYPE.ITEM_TYPE_TEXT.ordinal) {
            TextViewHolder(mInflater.inflate(R.layout.item_joke_text, parent, false))
        } else if (viewType == ITEM_TYPE.ITEM_TYPE_GIF.ordinal) {
            GifViewHolder(mInflater.inflate(R.layout.item_joke_gif, parent, false))
        } else if (viewType == ITEM_TYPE.ITEM_TYPE_VIDEO.ordinal) {
            VideoViewHolder(mInflater.inflate(R.layout.item_joke_vidio, parent, false))
        } else {
            null
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val joke = lists[position]
        var binding:ViewDataBinding?=null
        //评论内容
        when (holder) {
            is TextViewHolder -> binding= DataBindingUtil.bind<ItemJokeTextBinding>(holder.itemView)!!
            is ImageViewHolder -> {
                binding= DataBindingUtil.bind<ItemJokeImageBinding>(holder.itemView)!!
                binding.setClickListener {
                    when(it.id){
                        R.id.tvLarge->{
                            val dialog = ImageDetailDialog(mContext, joke.image!!.big!![0])
                            dialog.show()
                        }
                        R.id.llForward-> NewsManager.get().showShare(joke.text!!, joke.image!!.thumbnail_small!![0], joke.share_url!!)
                    }
                }
            }
            is GifViewHolder -> {
                binding= DataBindingUtil.bind<ItemJokeGifBinding>(holder.itemView)!!
                val url = joke.gif!!.images!![0]
                val imageLoadingPage = ImageLoadingPage(mContext)
                imageLoadingPage.layoutParams = FrameLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
                binding.llGif.removeAllViews()
                binding.llGif.addView(imageLoadingPage)
                binding.setClickListener {
                    when(it.id){
                        R.id.llGif->{
                            val dialog = ImageDetailDialog(mContext, joke.gif!!.images!![0])
                            dialog.show()
                        }
                        R.id.llForward->NewsManager.get().showShare(joke.text!!, joke.gif!!.gif_thumbnail!![0], joke.share_url!!)
                    }
                }
                imageLoadingPage.setImageUrl(url)

            }
            is VideoViewHolder -> {
                binding= DataBindingUtil.bind<ItemJokeVidioBinding>(holder.itemView)!!
                binding.jcPlayer.setUp(joke.video!!.video!![0], JCVideoPlayer.SCREEN_LAYOUT_LIST, joke.text!!)
                binding.setClickListener {
                    when(it.id){
                        R.id.llForward->NewsManager.get().showShare(joke.text!!, joke.video!!.thumbnail!![0],joke.share_url!!)
                    }
                }
                GlideUtils.load(mContext, joke.video!!.thumbnail!![0], binding.jcPlayer.thumbImageView)
            }
        }
        binding!!.setVariable(BR.joke,joke)
        addComment(joke,holder,binding)
    }

    private fun addComment(joke:Joke,holder:RecyclerView.ViewHolder,binding:ViewDataBinding){
        val top_comments = joke.top_comments
        if (top_comments != null && top_comments.size != 0) {
            if(holder is TextViewHolder )
                (binding as ItemJokeTextBinding).bottom!!.topComments.removeAllViews()
            else if(holder is GifViewHolder)
                (binding as ItemJokeGifBinding).bottom!!.topComments.removeAllViews()
            else if(holder is VideoViewHolder)
                (binding as ItemJokeVidioBinding).bottom!!.topComments.removeAllViews()
            for (comment in top_comments) {
                //行
                val linearLayout = LinearLayout(mContext)
                val layoutParams1 = LinearLayout.LayoutParams(
                        LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT)
                val layoutParams = LinearLayout.LayoutParams(
                        LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT)
                linearLayout.layoutParams = layoutParams1
                linearLayout.orientation = LinearLayout.HORIZONTAL
                linearLayout.setPadding(5, 2, 5, 2)

                //用户名
                val tv_name = TextView(mContext)
                tv_name.layoutParams = layoutParams
                tv_name.text = comment.u!!.name!! + ": "
                tv_name.setTextColor(Color.parseColor("#ff33b5e5"))

                //评论内容
                val tv_content = TextView(mContext)
                tv_name.layoutParams = layoutParams
                tv_content.text = comment.content
                tv_content.setTextColor(Color.BLACK)

                linearLayout.addView(tv_name)
                linearLayout.addView(tv_content)
                linearLayout.setBackgroundColor(Color.parseColor("#ffd7d7d7"))

                if(holder is TextViewHolder )
                    (binding as ItemJokeTextBinding).bottom!!.topComments.addView(linearLayout)
                else if(holder is GifViewHolder)
                    (binding as ItemJokeGifBinding).bottom!!.topComments.addView(linearLayout)
                else if(holder is VideoViewHolder)
                    (binding as ItemJokeVidioBinding).bottom!!.topComments.addView(linearLayout)
            }
        }
    }

    override fun getItemViewType(position: Int): Int {
        var ordinal = 0
        when (lists[position].type) {
            "text" -> ordinal = ITEM_TYPE.ITEM_TYPE_TEXT.ordinal   //序号
            "image" -> ordinal = ITEM_TYPE.ITEM_TYPE_IMAGE.ordinal
            "gif" -> ordinal = ITEM_TYPE.ITEM_TYPE_GIF.ordinal
            "video" -> ordinal = ITEM_TYPE.ITEM_TYPE_VIDEO.ordinal
        }
        return ordinal
    }

    override fun getItemCount(): Int {
        return if (lists == null) 0 else lists.size
    }


    class TextViewHolder(itemView: View) : BaseHolder(itemView)

    class ImageViewHolder(itemView: View) : BaseHolder(itemView) {

    }

    class GifViewHolder(itemView: View) : BaseHolder(itemView) {
    }

    class VideoViewHolder(itemView: View) : BaseHolder(itemView) {
    }

    open class BaseHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
    }

}